IntelliJ Platform Gradle Plugin: JetBrains Runtime (#1306)
JetBrains/intellij-sdk-docs: topics/appendix/tools/intellij_platform_gradle_plugin/tools_intellij_platform_gradle_plugin_jetbrains_runtime.md

<!-- Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license. -->

# JetBrains Runtime (JBR)

<link-summary>The IntelliJ Platform Gradle Plugin requires JetBrains Runtime (JBR), which can be provided in multiple ways.</link-summary>

<include from="tools_intellij_platform_gradle_plugin.md" element-id="Beta_Status"/>
<include from="tools_intellij_platform_gradle_plugin.md" element-id="faq"/>

> The IntelliJ Platform Gradle Plugin requires JetBrains Runtime (JBR) for running IDE locally, testing, or performing other actions that require Java Runtime adjusted by JetBrains.

Reviewer: what does "adjusted by JB" mean

Reviewer 2: I would start with a sentence or two of explanation what JBR is and why it exists. Then that IPGP needs it too.

> For the best experience, it is recommended to rely on JetBrains Runtime bundled with the IntelliJ Platform used for development.

Reviewer 2: ```suggestion
For the best experience, it is recommended to rely on JBR bundled with the IntelliJ Platform used for development.
```
Let's use "JBR" abbreviation everywhere after introducing it in the first paragraph.

> If one is not available, i.e., when targeting IntelliJ Platform snapshot or nightly releases, there are other ways available for providing JetBrains Runtime in a required version or variant.

Reviewer: Since when do snapshot releases not contain JBR?

Author: I don't exactly remember, but 2022.3.3 already doesn't have it.

> ## Bundled (default, recommended)

Reviewer 2: ```suggestion
## Bundled (Default)
```
And I would finish the section with "Using the bundled JBR is the recommended approach".

> IDE releases provided with JetBrains Download CDN are OS-specific and contain JetBrains Runtime (JBR) already bundled within the archive.

Reviewer: what is "JetBrains Download CDN"

This is the recommended way for developing plugin, as it comes in exactly the same version as JetBrains IDEs delivered to users.

To use this approach, no extra steps are required but declare the IntelliJ Platform, see: [](tools_intellij_platform_gradle_plugin_dependencies_extension.md#target-platforms).

## Obtained with IntelliJ Platform from Maven

> If the IntelliJ Platform is resolved with the IntelliJ Maven Repository instead of JetBrains Download CDN, it has no JetBrains Runtime (JBR) bundled.

Reviewer: "IntelliJ Maven Repository instead of JetBrains Download CDN" is unclear

> In such a case, it is necessary to provide it in the suitable version with the version read from the <path>[intellijPlatform]/dependencies.txt</path> file.

Reviewer 2: `[intellijPlatform]/dependencies.txt` is unclear to me. Is it the IDE home directory? Maybe an example could help?

Author: It's the IDE home directory but in the most of cases, it's long, ugly and should not be directly referred.

This can be easily achieved by using the `jetbrainsRuntime()` repository helper pointing to [JetBrains Runtime GitHub Releases](https://github.com/JetBrains/JetBrainsRuntime/releases/) and dependency helper with no arguments provided:

```kotlin
repositories {
  mavenCentral()
  intellijPlatform {
    defaultRepositories()
    jetbrainsRuntime()
    // ...
  }
}

dependencies {
  intellijPlatform {
    intellijIdeaCommunity("%ijPlatform%")

    jetbrainsRuntime()

    // ...
  }
}
```

> ## Declared explicitly

Reviewer 2: ```suggestion
## Declared Explicitly
```
Title Case.

It is possible to explicitly specify JetBrains Runtime version, variant, or exact build with:

```kotlin
repositories {
  mavenCentral()
  intellijPlatform {
    defaultRepositories()
    jetbrainsRuntime()
    // ...
  }
}

dependencies {
  intellijPlatform {
    intellijIdeaCommunity("%ijPlatform%")

    jetbrainsRuntime(version, variant, architecture)
    // or
    jetbrainsRuntimeExplicit(explicitVersion)

    // ...
  }
}
```

Provided `version`, `variant`, and `architecture` parameters along with the `explicitVersion` are used to resolve the JetBrains Runtime archives published on [GitHub Releases](https://github.com/JetBrains/JetBrainsRuntime/releases/) page.

> To correctly understand the pattern, refer to the archive names in format `jbr_[prefix]-[jdk]-[os]-[arch]-b[build].tar.gz`, such as `jbr_jcef-21.0.3-osx-aarch64-b446.1.tar.gz`, which can be split into the following parts:

Reviewer 2: ````suggestion
To correctly understand the pattern, refer to the archive names in format:
```
jbr_[prefix]-[jdk]-[os]-[arch]-b[build].tar.gz
```
e.g.:

```
jbr_jcef-21.0.3-osx-aarch64-b446.1.tar.gz
```
which can be split into the following parts:
````
Please avoid breaking the line when explaining the pattern.

- `prefix` = `jcef`
- `jdk` = `21.0.3`
- `os` = `osx`
- `arch` = `aarch64`
- `build` = `446.1`

Note that the `version` parameter is actually a combination of JDK and build numbers in format `[jdk]b[build]`.

To resolve the `jbr_jcef-21.0.3-osx-aarch64-b446.1.tar.gz` archive, you can use:
- `jetbrainsRuntime("21.0.3b446.1", "osx", "aarch64")`
- `jetbrainsRuntime("21.0.3b446.1")` (`os` and `architecture` are by default resolved with the current environment)
- `jetbrainsRuntimeExplicit("jbr_jcef-21.0.3-osx-aarch64-b446.1")`

See the [JetBrains Runtime releases page](https://github.com/JetBrains/JetBrainsRuntime/releases) for the list of available releases.

## Foojay Toolchains Plugin

> Since Gradle `8.4`, it is possible to specify JetBrains as a known JVM vendor and instruct Gradle to search for JetBrains Runtime within available toolchains (see [Toolchains for JVM projects](https://docs.gradle.org/current/userguide/toolchains.html)).

Reviewer: How does that differ from above? What is the use case? Which one should users prefer?

Author: It's explained below:

> Please note that the latest available JetBrains Runtime release is always resolved, which may lead to unexpected behaviors.
